### PR TITLE
[SPARK-43975][SQL] DataSource V2: Handle UPDATE commands for group-based sources

### PR DESCRIPTION
**Pull request by aokolnychyi**
### What changes were proposed in this pull request?

This PR adds support for group-based data sources in `RewriteUpdateTable` and builds on top of the existing logic.

The UPDATE command for group-based is executed differently based on whether there is a subquery in the condition. If there is no subquery, there is simply a projection to update matching rows. If there is a subquery, the rule would assign a union of two scans: one to find matching rows and one to carry over unmatched records in matched groups. The union is required so that subqueries can be rewritten as joins via `RewritePredicateSubquery`. In the future, we may provide a more optimized execution if the subquery is small. That said, the current approach would scale better and should be the default.

### Why are the changes needed?

These changes are needed per SPIP SPARK-35801.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR comes with tests. There are more tests in `AlignUpdateAssignmentsSuite`, which was merged earlier.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @aokolnychyi .

File: sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.spark.sql.catalyst.expressions.DynamicPruningExpression
+import org.apache.spark.sql.execution.{InSubqueryExec, ReusedSubqueryExec}
+import org.apache.spark.sql.execution.datasources.v2.BatchScanExec
+import org.apache.spark.sql.internal.SQLConf
+
+class GroupBasedUpdateTableSuite extends UpdateTableSuiteBase {
+
+  import testImplicits._
+
+  test("update runtime group filtering (DPP enabled)") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala:30]: Like the following, please merge two test case to remove repetition.
https://github.com/apache/spark/blob/fff0b71d24bf179ccf4f9fcd7a095dcafe19a77b/sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala#L1162-L1163

File: sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  }
+
+  test("update runtime group filtering (DPP disabled)") {
+    withSQLConf(SQLConf.DYNAMIC_PARTITION_PRUNING_ENABLED.key -> "false") {
+      checkUpdateRuntimeGroupFiltering()
+    }
+  }
+
+  test("update runtime group filtering (AQE enabled)") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala:42]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+      checkAnswer(
+        sql(s"SELECT * FROM $tableNameAsString"),
+        Row(1, -1, "hr") :: Row(2, 150, "software") :: Row(3, 120, "hr") :: Nil)
+
+      checkReplacedPartitions(Seq("hr"))
+    }
+  }
+
+  test("update runtime filter subquery reuse (AQE enabled)") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala:78]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/connector/UpdateTableSuiteBase.scala [hunk 4]
```diff
@@ -114,7 +114,7 @@ abstract class UpdateTableSuiteBase extends RowLevelOperationSuiteBase {
       Row(1, -1, "hr") :: Row(2, -1, "hardware") :: Row(3, -1, "hr") :: Nil)
   }
 
-  test("update with NULL conditions") {
+  test("update with NULL conditions on partition columns") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/UpdateTableSuiteBase.scala:117]: Thank you for clarifying this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala [hunk 5]
```diff
@@ -88,7 +167,7 @@ object RewriteUpdateTable extends RewriteRowLevelCommand {
   }
 
   // this method assumes the assignments have been already aligned before
-  private def buildUpdateProjection(
+  private def buildWriteDeltaUpdateProjection(
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala:170]: Nice renaming.

**Review by dongjoon-hyun — COMMENTED** [review]
I left a first-round review with minor comments about test cases. The other looks good to me, @aokolnychyi .

File: sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala (hunk 1 above)
Comment (aokolnychyi) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/GroupBasedUpdateTableSuite.scala]: Sounds good, I'll update DELETE and MERGE tests in a separate PR for consistency.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for updating.

**Comment by dongjoon-hyun** [comment]
All tests passed. Merged to master for Apache Spark 3.5.0.

**Comment by aokolnychyi** [comment]
Thank you, @dongjoon-hyun!

**Pull request closed without merging**